### PR TITLE
GSoC 2024: Internationalization Implementation for Zimit Frontend

### LINKED ISSUE #46: Interface should be multilingual
As part of the zimit 2.0 project and as a requirement from the project's sponsor, we need to make the youzim.it page available in farsi.

The number of strings is fairly limited and I can probably find a helpful native speaker to review or even do translations, but is there any specific process?

### PR DESCRIPTION
**Pull request by Onyx2406**
Fixes #46 
See also https://github.com/openzim/overview/issues/28

## Overview
This pull request introduces comprehensive internationalization (i18n) support to the Zimit frontend.

## What Has Been Done
- **Vue I18n Integration**: Implemented Vue I18n for managing translations within the application, with support for lazy-loaded language files.
- **Language Selection**: Added a language selection dropdown in the navbar, enabling users to switch languages dynamically. This feature is backed by automatic detection of the user's browser language, setting the application's default language accordingly.
- **English and French Translations**: Currently, the application supports English and French, with key UI elements and prompts available in both languages.

## What Is Left
- **TranslateWiki Integration**: Formal integration with TranslateWiki is pending.
- **Automated Browser Language Detection Improvement**: While initial support for detecting the browser's language is implemented, further testing and refinements is needed across different browsers and user configurations.
- **Preparation for TranslateWiki**: Organizing and documenting the translation file structure to facilitate future integration with TranslateWiki.
- **Expand Language Support**: Adding more languages based on community demand and contributions.

## Screenshots
image

### CODE REVIEW

**Comment by kelson42** [comment]
@Onyx2406 Thank you for your PR, here a few questions:
* Does it fix https://github.com/openzim/zimit-frontend/issues/46
* Please remove any code related to https://github.com/openzim/zimit-frontend/issues/28 and make an other PR.
* I would recommend to read https://github.com/openzim/overview/wiki/openZIM-code-contribution-workflow (if not already done)
* I general, for complex issues (like this one) we recommend to put a comment first explaining how you plan to solve the issue (this avoid to create large gap between expectation and PR)
* I have assigned a reviewer who will have a look during next week

**Comment by Onyx2406** [comment]
> @Onyx2406 Thank you for your PR, here a few questions:
> 
> * Does it fix [Interface should be multilingual #46](https://github.com/openzim/zimit-frontend/issues/46)
> * Please remove any code related to [Blacklist requests that are duplicates of existing resources or bound to fail #28](https://github.com/openzim/zimit-frontend/issues/28) and make an other PR.
> * I would recommend to read https://github.com/openzim/overview/wiki/openZIM-code-contribution-workflow (if not already done)
> * I general, for complex issues (like this one) we recommend to put a comment first explaining how you plan to solve the issue (this avoid to create large gap between expectation and PR)
> * I have assigned a reviewer who will have a look during next week

It does fix #46, but this PR only adds translations to the strings within this repository using Vue I18n. It doesn't translate the form details from Zimfarm. This PR only contains code to solve #46 and not #28. (I wanted to refer to https://github.com/openzim/overview/issues/28).

**Comment by kelson42** [comment]
@nikerabbit @Onyx2406 Do you know if this translation file can be indgested by Translatewiki?

**Comment by Onyx2406** [comment]
@kelson42, they are using a similar structure at https://github.com/mathjax/MathJax-i18n, so it should work.

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thank you; it's promising.

- We need RTL support
- I see that `human_size_limit` and `human_time_limit` are not i18n
- Languages list if static at the moment: requiring manually modifying two files to add a new locale. With translatewiki adding more all the time, we'd want something automatic

FYI (but it's a different PR) I believe there might be some strings in the backend (error message?) that would also require integration to completely close this

File: ui/src/locales/en.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{
+    "alertFeedback": {
```
Comment (benoit74) [inline ui/src/locales/en.json:2]: What is it used for?

File: ui/src/locales/en.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        "failureReasons": "A number of reasons can lead to a ziming failure. Most of the time, it's an inadequate URL… Please triple check it and create a new request. If that doesn't work, open a ticket in github.",
+        "success": "Success!",
+        "linkExpires": "The link below will expire and the file will be deleted after a week.",
+        "download": "Download",
+        "limitHit": "You have reached the maximum file size or duration allowed for free crawling. Contact us to help us purchase additional server space for you.",
+        "beingProcessed": "Your request is being processed",
+        "serverConverting": "One of our servers is currently converting that URL into a nice ZIM file. Depending on the number of pages and resources available, it can be a matter of minutes, hours or even days! Please be patient.",
+        "emailNotification": "You can close this window. You will get an email notification when the task is completed."
+    },
+    "switchButton": {
```
Comment (benoit74) [inline ui/src/locales/en.json:50]: What is it used for?

File: ui/src/locales/en.json [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        "missingContent": "The Zim file is incomplete or smaller than the original website",
+        "missingContentDesc": "Because of the very nature of this tool, we can’t leave it open for unlimited requests towards any website. That could be harmful both for our infrastructure, but also for the target websites. We currently enforce two limits: file size and time.",
+        "gotError": "I got an error message (no zim) or could not read a zim file",
+        "gotErrorDesc": "Triple-check the URL you entered, and if it is still not working then open a bug ticket on github. Indicate the target website, the request number (it’s in the email you received), and the device you tried to open your zim file on."
+    },
+    "loading": {
+        "loading": "Loading…"
+    },
+    "navbar": {
+        "logoAlt": "Youzim.it logo"
```
Comment (benoit74) [inline ui/src/locales/en.json:22]: This has not been used in code while it should

**Review by benoit74 — CHANGES_REQUESTED** [review]
Very good job, I'm happy to see we are already quite close to have completed this task.

I think you've forgot to i18n the footer as well.

I propose to split the integration with TranslateWiki to another PR, we would already benefit from merging this PR once concerns have been addressed (Right To Left support is very important, I don't know if it works as of today, please test and confirm/adapt). @rgaudin are you ok with it?

Regarding @rgaudin comment about the automatic detection of new languages added, I do not think it is mandatory, especially it should not break lazy loading of translations. I do not mind to have to edit one more file everytime a new language is added (I imagine TranslateWiki folks might even be able to do it themselves when they add new languages).

Would you mind to create the "secret" `qqq` language where TranslateWiki expects to have translation instructions in English?

> TranslateWiki Integration: Formal integration with TranslateWiki is pending.

I will take care of that once PR is merged

> Automated Browser Language Detection Improvement: While initial support for detecting the browser's language is implemented, further testing and refinements is needed across different browsers and user configurations.

Do you have something special in mind to be tested? Or is it just "normal" testing?

> Preparation for TranslateWiki: Organizing and documenting the translation file structure to facilitate future integration with TranslateWiki.

I think the only thing left is to add the `qqq` magic language with translation instructions. Do you have something else in mind?

> FYI (but it's a different PR) I believe there might be some strings in the backend (error message?) that would also require integration to completely close this

Definitely a second step, let's finish this first one for now.

**Comment by Nikerabbit** [comment]
> @Nikerabbit @Onyx2406 Do you know if this translation file can be indgested by Translatewiki?

Yes, we do support nested JSON. Also FYI we have recently updated our guidance for new projects: https://translatewiki.net/wiki/Translating:New_project

File: ui/src/locales/en.json (hunk 2 above)
Comment (Onyx2406) [reply ui/src/locales/en.json:50]: It's not used anywhere currently, I will remove unnecessary strings.

File: ui/src/locales/en.json (hunk 3 above)
Comment (Onyx2406) [reply ui/src/locales/en.json:22]: I will push the commits with these changes,

**Comment by Onyx2406** [comment]
Regarding splitting the integration with TranslateWiki into another PR, I completely agree. I have added the footer i18n.

I'll ensure that RTL support is tested and functional before merging. However, as the application currently supports only two languages which don't require RTL support, We can add RTL support to a future PR when additional RTL languages are added?

I've added the qqq language file for translation instructions.

For the automated browser language detection improvement, I'll do further testing across different browsers, so it's a normal testing and nothing special.

**Comment by benoit74** [comment]
Thank you ! Regarding RTL support, I do not mind to split it into another PR if there is too much work. It is however a pre-requisite for TranslateWiki integration and mostly a requirement for us, we have to translate the website into Farsi / Persian and already have persons ready to do the translation, so it will be in use quite quickly after TranslateWiki integration. Doesn't need to be perfect, but at least it should look pretty (e.g. I don't think we necessarily have to transfer all left buttons to the right and vice-versa, but at least the text should be readable and preferably right aligned when it was left aligned).

**Comment by Onyx2406** [comment]
Alright. In that case, it's better to push commits for RTL Support in this PR itself.

File: ui/src/components/NavBar.vue [hunk 4]
```diff
@@ -7,6 +7,7 @@
     <select @change="changeLanguage($event)" class="language-selector">
       <option value="en">English</option>
       <option value="fr">Français</option>
+      <option value="pers">Persian</option>
```
Comment (benoit74) [inline ui/src/components/NavBar.vue:10]: Value should by ISO639 so `far`, and label should be in Farsi so `فارسی`

Comment (benoit74) [reply ui/src/components/NavBar.vue:10]: Nota: this means you have to update the file name ^^

**Comment by benoit74** [comment]
Btw, do not forget to have a look at `human_size_limit` and `human_time_limit`, this must be internationalized as well (especially for RTL I think). 

Regarding the persian translation, is it a language you speak or did you just performed an automated translation? If it is an automated translation, we will definitely need someone else to review it, I can't read it at all.

**Comment by Nikerabbit** [comment]
FYI translatewiki.net prefers two letter language codes if one exists. Else we need to add code mapping.

**Comment by Onyx2406** [comment]
> just performed an automated translation

I had just performed an automated translation so we need someone to review it.

**Comment by Onyx2406** [comment]
Do we need to change the logo and buttons in RTL like this:

image

**Comment by benoit74** [comment]
@Onyx2406 I think this is exactly what we where referring to when mentioning adding support for RTL languages. 

@Popolechien @kelson42 @rgaudin do you confirm we want to reverse the layout when selecting a RTL language for Zimit as proposed in https://github.com/openzim/zimit-frontend/pull/51#issuecomment-1992289575?

Note that other screen(s) will have to be adapted as well.

**Comment by rgaudin** [comment]
> @Popolechien @kelson42 @rgaudin do you confirm we want to reverse the layout when selecting a RTL language for Zimit 

Yes

**Comment by Onyx2406** [comment]
> Note that other screen(s) will have to be adapted as well.

@benoit74 Any screen you are referring to other than the main page? RTL support (reverses the layout) now works well for Farsi. It would be great if you could test it.

**Comment by benoit74** [comment]
> @benoit74 Any screen you are referring to other than the main page? RTL support (reverses the layout) now works well for Farsi. It would be great if you could test it.

I was referring to the fact that 3 components (and their subcomponents) have to be adapted: Request, NotFound and  NewRequest.

Please tell me once code is ready for a new review (or even better, request it directly through the dedicated button if you can, button is supposed to be in the upper right "Reviewers" box, I never recall if external contributors have access to it or not).

**Review by benoit74 — CHANGES_REQUESTED** [review]
Thank you ! Still more stuff to fix unfortunately.

I don't get why `units.sizeLimit` is duplicated at `request.humanSizeLimit` and `units.timeLimit` is duplicated as `request.humanTimeLimit`. I suggest to keep only one at this stage (and prefer the `units` prefix which clearly indicates this is used in multiple screens).

There are some alignments issues on RTL:

image

image

image

image

Etc ... I think you got the point. Or is it due to the fact that my browser is not configured for a RTL language? I would prefer if we do not depend on this kind of browser setting.

The footer is not in proper order in RTL: Mozilla should be before - to the right - of Mozilla Open-source support award

image

Links in the FAQ are gone.

Pluralization is not handle appropriately (it wasn't handle before as well, but better to fix it now that I saw it):

image

I also realize we forgot to translate the email (in templates folder). Maybe we can do it in a second PR since this is backend code, to be done in Python. This makes me wonder how TranslateWiki will handle this, we will have to check with them.

File: ui/src/locales/en.json [hunk 5]
```diff
@@ -1,66 +1,67 @@
...
+        "advancedOptions": "advanced options"
     },
-    "notFound": {
+    "notFound": 
+    {
         "heading": "Not Found",
         "description": "Sorry, this content was not found. Maybe you'd like to go back."
     },
-    "request": {
-        "pending": "Pending",
```
Comment (benoit74) [inline ui/src/locales/en.json]: Why did you removed all these translations?

File: ui/src/locales/en.json [hunk 6]
```diff
@@ -1,66 +1,67 @@
...
-    "newRequest": {
+    "newRequest": 
+    {
         "heading": "Want an offline version of a website? Just",
         "boldZim": "Zim it",
         "urlPlaceholder": "Full URL of the website to convert",
         "emailPlaceholder": "Your e-mail to receive a download link. Address not kept",
         "submit": "Let's Zim it!",
-        "advancedOptions": "advanced options",
-        "true": "True",
```
Comment (benoit74) [inline ui/src/locales/en.json]: Why did you removed these translations?

File: ui/src/locales/fa.json [hunk 7]
```diff
@@ -1,78 +1,78 @@
 {
-    "faq": {
+    "faq": 
+    {
```
Comment (benoit74) [inline ui/src/locales/fa.json:3]: Why all in a sudden all `{` are moved to the next line? This is not standard JSON formatting, please revert (you should have `"faq": {` as before, and so on

**Review by benoit74 — CHANGES_REQUESTED** [review]
Thank you, we are getting closer but you broke few things in your last commits (see conversations)

Please do not spend too much time fine-tuning the exact words of French and Farsi translations, since you've made them with an automated translation tool we will not merge them anyway, they will have to be translated by native speakers on TranslateWiki.

Other topics to address:
- you updated all locales but not the `qqq.json`, while it is probably the most important one
- why did you sometimes place HTML code inside the text to translate (e.g. `app.thanksToMozilla`) and other times you split the text is many substrings (e.g `faq.gotErrorDesc1` and `faq.gotErrorDesc2`) ? I believe it would be way simpler in most situations (e.g. not when HTML code is too complex, but in every other situations) to always have HTML code inside the translation and one single string to translate, it will make the translator work way easier rather than being presented with only few words out of any context, or even a dot or a comma to not forget at the beginning of a translation
  - e.g you should have only one `footer` string (`app` is a bit vague) and its value should be `Powered by <a target="_blank" href="https://kiwix.org">Kiwix</a> and <a target="_blank" href="https://webrecorder.net">Webrecorder</a>, thanks to a <a target="_blank" href="https://www.mozilla.org/moss/">Mozilla Open-Source Support</a> Award`
- limits in Request.vue JS code are still not handled properly:
  - they are not translated like in the Faq.vue
  - pluralization is not handled
- "+" buttons of the FAQ are still not on the left when RTL language is used
  - please have a look, if it is too complex I do not mind to move this to an issue for later fix, it is not a big usability concern

File: ui/src/components/FaqEntry.vue [hunk 8]
```diff
@@ -1,41 +1,54 @@
...
-    text-align: left;
+    text-align: start;
   }
 
   .card-header .plus, .card-header .minus {
     position: absolute;
     top: .5em;
+  }
+
+  /* LTR specific styles */
```
Comment (benoit74) [inline ui/src/components/FaqEntry.vue:36]: Looks like it is possible to simplify by specifying both `left` and `right` value in all cases (ltr and rtl). See https://developer.mozilla.org/en-US/docs/Web/CSS/position?v=example#types_of_positioning and confirm if it works as expected on most browsers.

> If both left and right are specified, left wins when [direction](https://developer.mozilla.org/en-US/docs/Web/CSS/direction) is ltr (English, horizontal Japanese, etc.) and right wins when [direction](https://developer.mozilla.org/en-US/docs/Web/CSS/direction) is rtl (Persian, Arabic, Hebrew, etc.).

File: ui/src/components/NewRequest.vue [hunk 9]
```diff
@@ -1,13 +1,13 @@
 <template>
   <div class="container">
-    <h1>Want an offline version of a website? Just <strong>Zim it</strong>!</h1>
+    <h1>{{$t('newRequest.heading')}} <strong>{{$t('newRequest.boldZim')}}</strong>!</h1>
```
Comment (benoit74) [inline ui/src/components/NewRequest.vue:3]: This should be on single "thing" to translate, not two

**Review by benoit74 — CHANGES_REQUESTED** [review]
Thank you, we are getting closer to merging this.

Still some comments in conversation plus few below.

- I just realized that the zimit logo was previously centered on the page while it is not left or right ; you must put it back to center (sorry to not have realized this before, I did not got it right)
- FYI, we will not integrate with TranslateWiki right away, I will open an issue with details on what still has to be done (but does not have to be addressed in this issue)

**Comment by benoit74** [comment]
New issue about what is left (not part of this issue / PR): https://github.com/openzim/zimit-frontend/issues/53

File: ui/src/components/FaqEntry.vue (hunk 8 above)
Comment (Onyx2406) [reply ui/src/components/FaqEntry.vue]: Simplified it, works as expected.

**Review by benoit74 — CHANGES_REQUESTED** [review]
I'm sorry, looks like we didn't understood us well regarding the centering of the logo. I wanted to center only the logo, but not move the language button.

Expected UI in English:

image

Expected UI in Farsi:

image

Rest is OK now

File: ui/src/components/NavBar.vue [hunk 10]
```diff
@@ -1,8 +1,10 @@
 <template>
-  <header>
-    <router-link :to="{ name: 'home' }"><img alt="$t('navbar.logoAlt')" src="../assets/ZIMIT_LOGO_RGB.svg"></router-link>
-    <Loading/>
-
+  <header class="header">
+    <router-link :to="{ name: 'home' }" class="logo">
+      <img alt="$t('navbar.logoAlt')" src="../assets/ZIMIT_LOGO_RGB.svg">
+    </router-link>
+    <Loading class="loading"/>
```
[the remark this answers is not in the document]
Comment (Onyx2406) [reply ui/src/components/NavBar.vue]: Removed it.

File: ui/src/components/NavBar.vue [hunk 11]
```diff
@@ -1,8 +1,10 @@
 <template>
-  <header>
-    <router-link :to="{ name: 'home' }"><img alt="$t('navbar.logoAlt')" src="../assets/ZIMIT_LOGO_RGB.svg"></router-link>
-    <Loading/>
-
+  <header class="header">
```
[the remark this answers is not in the document]
Comment (Onyx2406) [reply ui/src/components/NavBar.vue]: Removed it.

**Comment by Onyx2406** [comment]
> Wouldn't it be simpler to move the select box out of the header, keep the header as it was and put an absolute positioning to the select box?

This worked, thanks!

File: ui/src/components/NavBar.vue [hunk 12]
```diff
@@ -1,24 +1,73 @@
...
+    margin-left: auto; 
+  }
+
+  .language-selector {
+    padding: .5rem 1rem;
+    border-radius: .25rem;
+    border: 1px solid #ced4da;
+    appearance: none; 
+    background-color: #fff;
+    background-image: url('data:image/svg+xml;charset=UTF8,<svg ...></svg>'); /* Add a custom dropdown arrow */
```
Comment (benoit74) [inline ui/src/components/NavBar.vue:63]: What is this? That's a strange SVG isn't it?

**Review by benoit74 — CHANGES_REQUESTED** [review]
One more thing ... Are you constantly generating code with a code generator? This is pretty annoying to spend so much time in reviews detecting code which means just nothing, or changes which are meaningless (random spaces added everywhere, ...)

**Comment by benoit74** [comment]
Thank you, unfortunately you again modified the indentation of many files.

I'm a bit sorry about that, but I will take over the end of this PR from this point, because I need to activate prettier and eslint on this, and this is way beyond the scope of this PR. 

Please consider that work is done from this point, I will finish what needs to be done. I will convert this to draft so that it is not merged by mistake.

Thank you.

**Comment by benoit74** [comment]
Done on our own finally

**Pull request closed without merging**